pkg/router/controller/route_secret_manager.go

```diff
@@ -0,0 +1,290 @@
+	authorizationv1 "k8s.io/api/authorization/v1"
+	kapi "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/authentication/user"
+	authorizationclient "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// routerServiceAccount is used to validate RBAC permissions for externalCertificate
+	routerServiceAccount = "system:serviceaccount:openshift-ingress:router"
+)
+
+// RouteSecretManager implements the router.Plugin interface to register
+// or unregister route with secretManger if externalCertificate is used.
+// It also reads the referenced secret to update in-memory tls.Certificate and tls.Key
+type RouteSecretManager struct {
+	// plugin is the next plugin in the chain.
+	plugin router.Plugin
+	// recorder is an interface for indicating route status.
+	recorder RouteStatusRecorder
+
+	secretManager secretmanager.SecretManager
+	secretsGetter corev1client.SecretsGetter
+	sarClient     authorizationclient.SubjectAccessReviewInterface
+}
+
+// NewRouteSecretManager creates a plugin wrapper that ....
+func NewRouteSecretManager(plugin router.Plugin, recorder RouteStatusRecorder, secretManager secretmanager.SecretManager, secretsGetter corev1client.SecretsGetter, sarClient authorizationclient.SubjectAccessReviewInterface) *RouteSecretManager {
+	return &RouteSecretManager{
+		plugin:        plugin,
+		recorder:      recorder,
+		secretManager: secretManager,
+		secretsGetter: secretsGetter,
+		sarClient:     sarClient,
+	}
+}
+
+func (p *RouteSecretManager) HandleNode(eventType watch.EventType, node *kapi.Node) error {
+	return p.plugin.HandleNode(eventType, node)
+}
+
+func (p *RouteSecretManager) HandleEndpoints(eventType watch.EventType, endpoints *kapi.Endpoints) error {
+	return p.plugin.HandleEndpoints(eventType, endpoints)
+}
+
+func (p *RouteSecretManager) HandleNamespaces(namespaces sets.String) error {
+	return p.plugin.HandleNamespaces(namespaces)
+}
+
+func (p *RouteSecretManager) Commit() error {
+	return p.plugin.Commit()
+}
+
+func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *routev1.Route) error {
+	klog.Info("Executing RouteSecretManager plugin...")
+
+	switch eventType {
+	case watch.Added:
+		// create new watch
+		if err := p.validateAndRegister(route); err != nil {
+			return err
+		}
+
+	case watch.Modified:
+		// remove old watch
+		if p.secretManager.IsRouteRegistered(route.Namespace, route.Name) {
+			if err := p.secretManager.UnregisterRoute(route.Namespace, route.Name); err != nil {
+				klog.Error("failed to unregister route", err)
+				return err
+			}
+		}
+		// create new watch
+		if err := p.validateAndRegister(route); err != nil {
+			return err
+		}
```

Comment:
It does seem odd that the plugin *always* unregisters and reregisters the route, even if the TLS configuration did not change.  In particular, I'm concerned about the overhead of validating and reregistering the route for updates that don't touch the TLS configuration.  If that is necessary to do, can you add a comment explaining why?  If it isn't necessary, please either add a check to skip the unnecessary deregister-reregister or add a to-do comment to improve the logic before GA.

Reply:
I got the validation and reregistration overhead concern here. The plugin *always* needs to unregister and reregister the route even if the TLS configuration did not change, for two reasons

1.  The `HandleRoute()` method does not carry the old route spec; it's only the new route that propagates in the plugin chain. So, there isn't a definite way to compare the old TLS configuration and the new TLS configuration, and decide whether there is any change happened or not. Hence, we assume that it's *always* updated, so reregistration is required.
2. Another reason for always doing reregistration is to have an updated `secretHandler`. If we don't always create a new handler (with every route update event), there may be changes of serving stale route spec in the next plugin chain, when the referenced `secret` is updated or deleted. Sending outdated route in the next plugin, may impact their expected functionality.

https://github.com/openshift/router/blob/c9ce8871adb319e79839566fd483f7d763c17610/pkg/router/controller/route_secret_manager.go#L163-L207

Reply:
If we want to improve the logic for GA, then I think we could rely on `secretManager` to give the previously referenced secret name. It could have a new method called `GetSecretName()` to return the previously registered secret name. And we will only unregister and re-register if it's updated.

This could look something like

```go
	case watch.Modified:

		newHasExt := hasExternalCertificate(route)
		oldHadExt := p.secretManager.IsRouteRegistered(route.Namespace, route.Name)

		if newHasExt { // new route has externalCertificate
			if oldHadExt { // old route was also having externalCertificate
				oldSecret := p.secretManager.GetSecretName(route.Namespace, route.Name)
				if oldSecret != route.Spec.TLS.ExternalCertificate.Name { // externalCertificate is updated
					// unregister and re-register
					p.secretManager.UnregisterRoute(route.Namespace, route.Name)
					p.validateAndRegister(route)
				} else {
					// do nothing if externalCertificate is not updated
				}
			} else { // old route was not having externalCertificate
				p.validateAndRegister(route)
			}
		} else { // new route doesn't have externalCertificate
			if oldHadExt { // old route was having externalCertificate
				p.secretManager.UnregisterRoute(route.Namespace, route.Name)
			} else {
				// do nothing : no update to externalCertificate
			}
		}
```

This could solve the issue of *always* doing unregister and re-register. 

Along with that, we also need to ensure that the `secretHandler` always processes the updated route object whenever there is a secret update or delete event.

I think this can be archived by getting the updated route object from the client. So the `UpdateFunc` and `DeleteFunc` of `secretHandler` will first query the apiserver for the latest route, and then will continue the next steps. 

Let me know if this approach would improve the overhead issue. We can update it for GA.
If there are any other suggestions, we can discuss them too.

Reply:
I've added a comment in the code for now, with a to-do to improve the logic while going to GA.

Placeholder to track : https://issues.redhat.com/browse/CFE-1057